File: day01/taxi.pl
#!/usr/bin/env perl
use strict;
use warnings;
use 5.24.0;

my %dirv = (N => [0,1],
            S => [0,-1],
            E => [1,0],
            W => [-1,0]
           );

my %r = (N => 'E',
         E => 'S',
         S => 'W',
         W => 'N'
        );

my %l = (N => 'W',
         E => 'N',
         S => 'E',
         W => 'S'
        );

my @pos = (0, 0);
my $dir = 'N';

while (<>) {
    chomp;
    my @moves = split /, /;
    for my $move (@moves) {
        my $turn = substr($move, 0, 1);
        my $n = substr($move, 1);

        if ($turn eq 'R') {
            $dir = $r{$dir};
        } else {
            $dir = $l{$dir};
        }
        my $dirv = $dirv{$dir};

        $pos[0] += $dirv->[0] * $n;
        $pos[1] += $dirv->[1] * $n;
    }
}

say "@pos";
say abs($pos[0]) + abs($pos[1]);

File: day01/taxi.py
#!/usr/bin/env python
from sys import argv
from numpy import *

dirv = {}
dirv['N'] = array([0, 1])
dirv['S'] = array([0,-1])
dirv['E'] = array([1, 0])
dirv['W'] = array([-1,0])

r = {}
r['N'] = 'E'
r['E'] = 'S'
r['S'] = 'W'
r['W'] = 'N'

l = {}
l['N'] = 'W'
l['E'] = 'N'
l['S'] = 'E'
l['W'] = 'S'

pos = array([0, 0])
dir = 'N'

script, filename = argv
with open(filename) as f :
    for line in f :
        line = line.rstrip("\n")
        moves = line.split(", ")
        for move in moves :
            turn = move[:1]
            n = int(move[1:])

            if turn == "R" :
                dir = r[dir]
            else :
                dir = l[dir]

            v = dirv[dir]
            pos = pos + v * n

print pos
print abs(pos[0]) + abs(pos[1])
